### PR TITLE
fix: preferences controller state sync

### PR DESCRIPTION
**Pull request by montelaidev**
## **Description**

This PR syncs the state of accounts between the AccountsController and PreferencesController to keep them in sync. 

The PreferencesController now mirrors the accounts from the AccountsController's InternalAccounts state. This ensures the PreferencesController always reflects the accurate account state.

Any write operations like setting the selected account or label are passed through to the AccountsController:

- `setSelectedAddress` calls `setSelectedAccount` on the AccountsController
- `setAccountLabel` calls `setAccountName` on the AccountsController

To get the selected account, the PreferencesController retrieves it from the AccountsController instead of internal state.

This allows the AccountsController to be the source of truth, while any write operations called on the PreferencesController are delegated to the AccountsController.

## **Related issues**

Fixes: https://github.com/MetaMask/accounts-planning/issues/445

## **Manual testing steps**

1. In the `develop` branch, create multiple accounts and update their names
2. Change to the code in this branch
3. Verify the state of accounts is consistent
4. Make sure basic operations like swap, send, and dapp connection is consistent with the accounts state

## **Screenshots/Recordings**

Not applicable

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/controllers/preferences.js [hunk 1]
```diff
@@ -386,76 +379,39 @@ export default class PreferencesController {
     return textDirection;
   }
 
-  /**
-   * Removes an address from state
-   *
-   * @param {string} address - A hex address
-   * @returns {string} the address that was removed
-   */
-  removeAddress(address) {
```
Comment (montelaidev) [inline app/scripts/controllers/preferences.js:395]: removeAddress was never used.

**Comment by montelaidev** [comment]
@metamask-bot update-policies

**Comment by montelaidev** [comment]
@metamaskbot update-policies

**Review by ccharly — DISMISSED** [review]

**Comment by montelaidev** [comment]
@metamaskbot update-policies

**Review by ccharly — DISMISSED** [review]

File: app/scripts/controllers/preferences.js [hunk 2]
```diff
@@ -387,76 +379,39 @@ export default class PreferencesController {
...
-    });
-    this.store.updateState({ identities });
-  }
-
   /**
    * Setter for the `selectedAddress` property
    *
-   * @param {string} _address - A new hex address for an account
+   * @deprecated - Please use setSelectedAccount from the AccountsController
+   * @param {string} address - A new hex address for an account
```
Comment (danroc) [inline app/scripts/controllers/preferences.js:386]:
```suggestion
   * @deprecated - Use setSelectedAccount from the AccountsController.
```

File: app/scripts/controllers/preferences.test.js [hunk 3]
```diff
@@ -76,74 +107,158 @@ describe('preferences controller', () => {
...
-            address: '0xda22le',
-          },
-          '0x7e57e2': {
-            name: 'Account 2',
-            address: '0x7e57e2',
+  describe('setAccountLabel', () => {
+    const mockName = 'mockName';
+    const firstAddress = '0x1f9090aaE28b8a3dCeaDf281B0F12828e676c326';
+    const secondAddress = '0x0affb0a96fbefaa97dce488dfd97512346cf3ab8';
+    it('updating name from preference controller will update the name in accounts controller and preferences controller', () => {
```
Comment (danroc) [inline app/scripts/controllers/preferences.test.js:114]:
```suggestion
    const secondAddress = '0x0affb0a96fbefaa97dce488dfd97512346cf3ab8';

    it('updating name from preference controller will update the name in accounts controller and preferences controller', () => {
```

File: app/scripts/controllers/preferences.test.js [hunk 4]
```diff
@@ -76,74 +107,158 @@ describe('preferences controller', () => {
...
-        identities: {
-          '0xda22le': {
-            name: 'Account 1',
-            address: '0xda22le',
-          },
-          '0x7e57e2': {
-            name: 'Account 2',
-            address: '0x7e57e2',
+  describe('setSelectedAddress', () => {
+    it('updating selectedAddress from preferences controller will update the selectedAccount in accounts controller and preferences controller', () => {
```
Comment (danroc) [inline app/scripts/controllers/preferences.test.js:203]:
```suggestion
    it('updating selectedAddress from preferences controller updates the selectedAccount in accounts controller and preferences controller', () => {
```

File: app/scripts/controllers/preferences.test.js [hunk 5]
```diff
@@ -76,74 +107,158 @@ describe('preferences controller', () => {
...
+
+      const { selectedAddress: updatedSelectedAddress } =
+        preferencesController.store.getState();
+
+      const updatedSelectedAccount = accountsController.getSelectedAccount();
+
+      expect(updatedSelectedAddress).toBe(updatedSelectedAccount.address);
+    });
+
+    it('updating selectedAccount from accounts controller should update the selectedAddress in preferences controller', () => {
```
Comment (danroc) [inline app/scripts/controllers/preferences.test.js:233]:
```suggestion
    it('updating selectedAccount from accounts controller updates the selectedAddress in preferences controller', () => {
```

File: app/scripts/controllers/preferences.js [hunk 6]
```diff
@@ -387,76 +379,39 @@ export default class PreferencesController {
...
-      identities[address] = { name: `Account ${identityCount + 1}`, address };
-    });
-    this.store.updateState({ identities });
-  }
-
   /**
    * Setter for the `selectedAddress` property
    *
-   * @param {string} _address - A new hex address for an account
+   * @deprecated - Please use setSelectedAccount from the AccountsController
```
Comment (danroc) [inline app/scripts/controllers/preferences.js:385]:
```suggestion
   * @deprecated - Use setSelectedAccount from the AccountsController.
```

File: app/scripts/controllers/preferences.js [hunk 7]
```diff
@@ -387,76 +379,39 @@ export default class PreferencesController {
...
+    this.messagingSystem.call(
+      'AccountsController:setSelectedAccount',
+      account.id,
+    );
   }
 
   /**
    * Getter for the `selectedAddress` property
    *
+   * @deprecated - Please use the getSelectedAccount from the AccountsController
```
Comment (danroc) [inline app/scripts/controllers/preferences.js:406]:
```suggestion
   * @deprecated - Use the getSelectedAccount from the AccountsController.
```

File: app/scripts/controllers/preferences.js [hunk 8]
```diff
@@ -471,21 +426,28 @@ export default class PreferencesController {
   /**
    * Sets a custom label for an account
    *
-   * @param {string} account - the account to set a label for
+   * @deprecated - Please use setAccountName from the AccountsController
```
Comment (danroc) [inline app/scripts/controllers/preferences.js:429]:
```suggestion
   * @deprecated - Use setAccountName from the AccountsController.
```

**Review by danroc — APPROVED** [review]

**Review by ccharly — APPROVED** [review]

**Pull request merged**